src/libmeasurement_kit/ooni/facebook_messenger.cpp

```diff
@@ -0,0 +1,221 @@
+        // skipping stun for now
+        if (service_and_ips.first == "stun") {
+            continue;
+        }
+        ips_count += service_and_ips.second.size();
+    }
+    if (ips_count == 0) {
+        cb(entry);
+        return;
+    }
+    Var<int> ips_tested(new int(0));
+
+    auto tcp_cb = [=](std::string service, std::string ip, int port) {
+        return [=](Error err, Var<net::Transport> txp) {
+            bool close_txp = true; // if connected, we must disconnect
+            if (!!err) {
+                logger->info("tcp failure to %s at %s:%d", service.c_str(),
+                    ip.c_str(), port);
+                (*entry)["facebook_" + service + "_blocking"] = true;
+                (*entry)["facebook_dns_blocking"] = true;
+                close_txp = false;
+            } else {
+                logger->info("tcp success to %s at %s:%d", service.c_str(),
+                    ip.c_str(), port);
+                (*entry)["facebook_" + service + "_blocking"] = false;
+            }
+            *ips_tested += 1;
+            if (ips_count == *ips_tested) {
+                if (close_txp == true) {
+                    txp->close([=] {
+                        cb(entry);
+                        return;
+                    });
+                } else {
+                    cb(entry);
+                    return;
+                }
+            } else {
+                if (close_txp == true) {
+                    // XXX optimistic closure
+                    txp->close([=] {});
+                }
+            }
+        };
+    };
+
+    for (auto const& service_and_ips : fb_service_ips) {
+        std::string service = service_and_ips.first;
+        // skipping stun for now
+        if (service == "stun") {
+            continue;
+        }
+        for (auto const& ip : service_and_ips.second) {
+            Settings tcp_options;
+            tcp_options["host"] = ip;
+            int port = 443; //XXX hardcoded
+            tcp_options["port"] = port;
+            tcp_options["net/timeout"] = 10.0; //XXX hardcoded
+            templates::tcp_connect(tcp_options, tcp_cb(service, ip, port),
+                                   reactor, logger);
+        }
+    }
+    return;
+}
+
+void facebook_messenger(Settings options,
+              Callback<Var<report::Entry>> callback,
+              Var<Reactor> reactor, Var<Logger> logger) {
+    logger->info("starting facebook_messenger");
+    Var<Entry> entry(new Entry);
+
+    std::map<std::string, std::string>
+        FB_SERVICE_HOSTNAMES = {
+            {"stun", "stun.fbsbx.com"},
+            {"b_api", "b-api.facebook.com"},
+            {"b_graph", "b-graph.facebook.com"},
+            {"edge", "edge-mqtt.facebook.com"},
+            {"external_cdn", "external.xx.fbcdn.net"},
+            {"scontent_cdn", "scontent.xx.fbcdn.net"},
+            {"star", "star.c10r.facebook.com"}};
+
+    options["geoip_asn_path"] = "GeoIPASNum.dat";
```

Comment:
You don't need this. With properly configured upstream options this file should already be set to point to the proper file.